[debugger][wasm] Implement Debugger.IsAttached on wasm

thaystg opened the pull request:
Debugger.IsAttached is now working on wasm. And can be used to detect if debugger is attached.

Fix #42411

thaystg committed:
Debugger.IsAttached is now working on wasm. And can be used to detect if debugger is attached.
Fix #42411

CoffeeFlux commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -53,6 +53,7 @@ EMSCRIPTEN_KEEPALIVE gboolean mono_wasm_get_array_values (int object_id, int sta
 EMSCRIPTEN_KEEPALIVE gboolean mono_wasm_invoke_getter_on_object (int object_id, const char* name);
 EMSCRIPTEN_KEEPALIVE gboolean mono_wasm_invoke_getter_on_value (void *value, MonoClass *klass, const char *name);
 EMSCRIPTEN_KEEPALIVE gboolean mono_wasm_get_deref_ptr_value (void *value_addr, MonoClass *klass);
+EMSCRIPTEN_KEEPALIVE void mono_wasm_set_is_debugger_attached (gboolean is_attached);
```
Is there a reason you have to forward declare this?

thaystg replied:
No, should I remove?

CoffeeFlux replied:
It doesn't harm anything, but I saw it's been done on a bunch of functions in this file that aren't called before the corresponding implementation and was wondering if I was missing something.

CoffeeFlux replied:
Ah no, there does seem to be a reason... Those are wrapped in `G_BEGIN_DECLS`, so I guess they exist so this correctly builds as C++ or something? Maybe someone else remembers what was the motivation for this.

CoffeeFlux commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -1817,6 +1823,9 @@ var MonoSupportLib = {
 		mono_wasm_get_loaded_files: function() {
+			if (!this.mono_wasm_set_is_debugger_attached)
+				this.mono_wasm_set_is_debugger_attached = Module.cwrap ('mono_wasm_set_is_debugger_attached', 'void', ['bool']);
+			this.mono_wasm_set_is_debugger_attached(true);
```
```suggestion
			this.mono_wasm_set_is_debugger_attached (true);
```

thaystg committed:
Update src/mono/wasm/runtime/library_mono.js

Co-authored-by: Ryan Lucia <ryan@luciaonline.net>

lewing reviewed (commented):
I think you want Target.detachedFromTarget instead of Target.targetDestroyed but I'm not sure we actually care about tracking detachment.

thaystg commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -167,6 +167,11 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, string meth
                         break;
                     }
 
+                case "Target.targetDestroyed":
```
The page is still opened, so the debug target is still there.
When we close the debugger devtools page we receive this message.

thaystg replied:
To test I added this method on program.cs that is called on wasm sample.

```
 [MethodImpl(MethodImplOptions.NoInlining)]
        public static string TestMeaning2()
        {
            return Debugger.IsAttached.ToString();
        }
```

And this method is called by the button that you will see in the video.

thaystg replied:
I think we care about tracking detachment because if we have a dynamic Assembly.Load we don't want to send information to the debugger because it's not attached anymore.

thaystg replied:
When the wasm webpage is closed we receive an Inspector.Detached message and this case I think it doesn't matter.

thaystg replied:
I don't know when we would receive Target.detachedFromTarget, I couldn't find it anywhere.

thaystg replied:
@lewing I will wait your review to merge! :)

thaystg commented:
[Uploading thays_implement_42411.zip…]()

thaystg commented:
[thays_implement_42411_2.zip](https://github.com/dotnet/runtime/files/5256805/thays_implement_42411_2.zip)

lewing commented:
This looks ok for adding the property but doesn't actually do anything with it.  I'd like to see the lazy loaded assembly events short circuit if the debugger isn't attached.

lewing reviewed (approved):
This looks ok as is but I'd like to see the value used on the js side as well to avoid doing extra work there.

thaystg committed:
Using the infrastructure to not send dynamically loaded assemblies if debugger is not attached.
Changing where to check if the assembly is already added to avoid unnecessary checks.

thaystg committed:
Merge branch 'thays_implement_42411' of github.com:thaystg/runtime into thays_implement_42411

* 'thays_implement_42411' of github.com:thaystg/runtime:
  Update src/mono/wasm/runtime/library_mono.js

thaystg commented:
Larry, I pushed the change to use the new infrastructure to avoid sending info if the debugger is not connected, and to send when the debugger is attached.

thaystg commented:
@lewing , I pushed the change to use the new infrastructure to avoid sending info if the debugger is not connected, and to send when the debugger is attached.

captainsafia commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -497,6 +500,15 @@ assembly_loaded (MonoProfiler *prof, MonoAssembly *assembly)
+		return;
+	}
+
+	if (mono_wasm_assembly_already_added(assembly->aname.name))
```
Won't this always return true even if an assembly hasn't been loaded into the DebugStore via the event? For example.

1. User launches an app without attaching to a debugger.
2. User navigates to a page that lazily-loads an assembly.
3. Assembly gets loaded and is added to the [assemblies list](https://github.com/dotnet/runtime/blob/ddc83ad2b6f0127f81486301185c57f8e9aac055/src/mono/wasm/runtime/driver.c#L185).
4. `mono_wasm_asm_loaded` event isn't fired since debugger isn't attached.
5. User attaches to a debugger.
6. `assembly_loaded` callback is invoked from `mono_wasm_set_is_debugger_attached`.
7. This check succeeds because the assembly was added in step 3.

Let me know if my understanding of this is incorrect.

thaystg replied:
Lazy loaded assemblies will be in the files that are passed to _handle_loaded_asset?

captainsafia replied:
They are added via `Assembly.Load` in [this code](https://github.com/dotnet/aspnetcore/blob/master/src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs#L113-L119). `_load_assets_and_runtime` is only used for assemblies that are loaded when the app is initialized.

thaystg replied:
But we only add things to [assemblies list](https://github.com/dotnet/runtime/blob/ddc83ad2b6f0127f81486301185c57f8e9aac055/src/mono/wasm/runtime/driver.c#L185), if the come from _load_assets_and_runtime, am I wrong?

captainsafia replied:
Nope! I had a gap in my understanding. This check will work fine. Resolving this comment now.

lewing commented on src/mono/wasm/runtime/driver.c:
```diff
@@ -206,15 +205,15 @@ mono_wasm_add_assembly (const char *name, const unsigned char *data, unsigned in
 	WasmAssembly *entry = assemblies;
 	while (entry != NULL) {
-		if (strcmp (entry->assembly.name, assembly_name) == 0)
+		if (strncmp (entry->assembly.name, assembly_name, strlen(entry->assembly.name - 4)) == 0)
```
hmmm isn't this is potentially fragile if the assemblies are MyAssembly.dll and MyAssembly.Browser.dll?  Should at least check that the lengths match as well.

lewing commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -1817,6 +1823,9 @@ var MonoSupportLib = {
 		mono_wasm_get_loaded_files: function() {
+			if (!this.mono_wasm_set_is_debugger_attached)
+				this.mono_wasm_set_is_debugger_attached = Module.cwrap ('mono_wasm_set_is_debugger_attached', 'void', ['bool']);
+			this.mono_wasm_set_is_debugger_attached (true);
```
There is no easy way to check the attached state from js now though right?

thaystg replied:
I searched if there is an easy way to get debugger state from js and I didn't find.
And another thing, if the debugger is attached but it's not using our ProxyDriver we don't want to send any information right?

lewing replied:
correct

thaystg committed:
Checking the assembly name size.
